client/src/main/java/io/pravega/client/stream/impl/Pinger.java

```diff
@@ -1,142 +0,0 @@
-/**
- * Copyright Pravega Authors.
- *
- * Licensed under the Apache License, Version 2.0 (the "License");
- * you may not use this file except in compliance with the License.
- * You may obtain a copy of the License at
- *
- *     http://www.apache.org/licenses/LICENSE-2.0
- *
- * Unless required by applicable law or agreed to in writing, software
- * distributed under the License is distributed on an "AS IS" BASIS,
- * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
- * See the License for the specific language governing permissions and
- * limitations under the License.
- */
-package io.pravega.client.stream.impl;
-
-import com.google.common.annotations.VisibleForTesting;
-import io.grpc.Status;
-import io.grpc.StatusRuntimeException;
-import io.pravega.client.control.impl.Controller;
-import io.pravega.client.stream.Stream;
-import io.pravega.client.stream.Transaction;
-import io.pravega.common.Exceptions;
-import java.util.HashSet;
-import java.util.Set;
-import java.util.UUID;
-import java.util.concurrent.ScheduledExecutorService;
-import java.util.concurrent.ScheduledFuture;
-import java.util.concurrent.TimeUnit;
-import java.util.concurrent.atomic.AtomicBoolean;
-import java.util.concurrent.atomic.AtomicReference;
-import javax.annotation.concurrent.GuardedBy;
-import lombok.AccessLevel;
-import lombok.Getter;
-import lombok.extern.slf4j.Slf4j;
-
-import static io.pravega.common.Exceptions.unwrap;
-
-/**
- * Pinger is used to send pings to renew the transaction lease for active transactions.
- * It invokes io.pravega.client.control.impl.Controller#pingTransaction() on the controller to renew the lease of active
- * transactions. Incase of a controller instance not being reachable the controller client takes care of retrying on a
- * different controller instance see io.pravega.client.control.impl.ControllerResolverFactory for details.
- */
-@Slf4j
-public class Pinger implements AutoCloseable {
```

Comment:
Deleting this breaks backwards compatibility and makes the interface worse. Currently the way the client works is that transactions are kept alive as long as the application's code is holding onto the transaction object in memory. When they drop the object it may timeout. This makes the pinging a purely internal concern, because an application cannot hold a TXN object and have no longer work. By removing the background process, the application is now responsible for doing something that they were not before. Of course this means that there isn't code for it in the application. So when an app upgrades, their transactions will suddenly start timing out when they didn't before. Additionally every application needs to implement the same logic that is in this class, or just switch to fixed timeouts and handle the case where the transaction timeout. I don't think this is a net improvement from a user's perspective.

Reply:
To maintain backward compatibility the existing pingTransaction API on Controller has been retained (& deprecated) and it will cause the transaction lease to get extended by that amount. Removing the pinger service in Client does not break backward compatibility as the Client always get upgraded after the Servers do.

Reply:
> Additionally every application needs to implement the same logic that is in this class, or just switch to fixed timeouts and handle the case where the transaction timeout. I don't think this is a net improvement from a user's perspective.

Post upgrade to the new design, user applications may need to adjust their lease periods once, but that is a one time change.  However, there are no regular "pings" required by the new design and hence there is no need for applications to implement these.

Reply:
If a user needs to change their code (which is where timeouts are specified) then it is not backward compatible, and hence needs to be behind a flag or config value of some sort so that existing apps keep working until changes are made.